### PR TITLE
rust: fix rust cross-compile

### PR DESCRIPTION
**Pull request by cleverca22**
reasoning:
sjlj (short jump long jump) exception handling makes no sense on x86_64, it's forcably slowing programs down as it produces a constant overhead. On x86_64 we have SEH (Structured Exception Handling) and we should use that. On i686, we do not have SEH, and have to use sjlj with dwarf2. Hence it's now conditional on x86_32

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by angerman** [comment]
refs
- https://stackoverflow.com/questions/15670169/what-is-difference-between-sjlj-vs-dwarf-vs-seh 
- https://github.com/Martchus/PKGBUILDs/issues/56

In the end building sjlj for x86_64 will just cause a forced performance penalty, as well as make building `rustc` impossible.

**Comment by arcnmx** [comment]
I believe `--disable-sjlj-exceptions` is necessary to cross compile rust on i686 - at least is with the binary releases. It might make sense to be able to customize the exception type in the platform config?

**Comment by angerman** [comment]
> I believe `--disable-sjlj-exceptions` is necessary to cross compile rust on i686 - at least is with the binary releases. It might make sense to be able to customize the exception type in the platform config?

i686 is 32bit and that’s where this patch leaves sjlj enabled. It only does not force enable sjlj on x86_64.

**Comment by Ericson2314** [comment]
Sorry didn't see this yet. I know @cleverca22 has been stuck with this far too long! reading now.

File: pkgs/development/compilers/gcc/common/configure-flags.nix [hunk 1]
```diff
@@ -58,8 +58,9 @@ let
       "--with-gnu-as"
       "--with-gnu-ld"
       "--disable-debug"
-      "--enable-sjlj-exceptions"
       "--disable-win32-registry"
+    ] ++ lib.optionals (crossMingw && targetPlatform.isx86_32) [
+      "--enable-sjlj-exceptions"
```
Comment (Ericson2314) [inline pkgs/development/compilers/gcc/common/configure-flags.nix:63]: Can you put some/all of the commit message as a comment here? My rule of thumb is the language that describes the (new) *state* should be comments, while the language that describes the transition should be the commit message.

File: pkgs/development/compilers/gcc/common/configure-flags.nix [hunk 2]
```diff
@@ -81,13 +82,14 @@ let
...
     ] ++ lib.optionals crossMingw [
-      "--enable-sjlj-exceptions"
       "--enable-hash-synchronization"
       "--enable-libssp"
       "--disable-nls"
-      "--with-dwarf2"
       # To keep ABI compatibility with upstream mingw-w64
       "--enable-fully-dynamic-string"
+    ] ++ lib.optionals (crossMingw && targetPlatform.isx86_32) [
+      "--enable-sjlj-exceptions"
```
Comment (Ericson2314) [inline pkgs/development/compilers/gcc/common/configure-flags.nix:91]: Does this need to be here twice? I don't think so---I think it is just an artifact of us having not cleaned up the list yet out of laziness/avoiding mass rebuilds.

File: pkgs/development/compilers/gcc/common/configure-flags.nix [hunk 3]
```diff
@@ -81,13 +82,14 @@ let
...
+      "--with-dwarf2"
```
Comment (Ericson2314) [inline pkgs/development/compilers/gcc/common/configure-flags.nix:92]: Can move up above, especially if the previous line isn't needed as I think.

**Review by Ericson2314 — APPROVED** [review]

**Comment by arcnmx** [comment]
> i686 is 32bit and that’s where this patch leaves sjlj enabled. It only does not force enable sjlj on x86_64.

Right, what I meant is that this change trades one hard-coded config for another, and that there are similar reasons one might want to change the i686 config flags in a similar way. It seems like it'd be useful to be able to use something like `crossSystem.platform.gcc.exceptionHandler = "sjlj"` so that the seh/sjlj/dwarf alternatives can be selected where they're needed.

File: pkgs/development/compilers/gcc/common/configure-flags.nix [hunk 4]
```diff
@@ -80,14 +98,6 @@ let
     ] ++ lib.optionals (targetPlatform.libc == "musl") [
       # musl at least, disable: https://git.buildroot.net/buildroot/commit/?id=873d4019f7fb00f6a80592224236b3ba7d657865
       "--disable-libmpx"
-    ] ++ lib.optionals crossMingw [
-      "--enable-sjlj-exceptions"
-      "--enable-hash-synchronization"
-      "--enable-libssp"
-      "--disable-nls"
```
Comment (js6pak) [inline pkgs/development/compilers/gcc/common/configure-flags.nix:87]: @angerman @Ericson2314 Were these flags moved to `crossStageStatic` (now called `withoutTargetLibc`) by accident?

Comment (js6pak) [reply pkgs/development/compilers/gcc/common/configure-flags.nix:87]: It was done in @angerman's commit here [`f2a33e4` (#89582)](https://github.com/NixOS/nixpkgs/pull/89582/commits/f2a33e4486b40434241b964be5c685600c84c3a7), with no comment

**Pull request merged**